cmake/modules/extensions.cmake

```diff
@@ -1583,17 +1594,45 @@ function(zephyr_build_string outvar)
     )
   endif()
 
+  if(DEFINED BUILD_STR_BOARD AND BUILD_STR_SOC)
+    message(FATAL_ERROR
+      "zephyr_build_string(${ARGV0} <list> ...) given BOARD and SOC:"
+      " only one of these can be provided"
+    )
+  endif()
+
   if(DEFINED BUILD_STR_BOARD_REVISION AND NOT BUILD_STR_BOARD)
     message(FATAL_ERROR
       "zephyr_build_string(${ARGV0} <list> BOARD_REVISION ${BUILD_STR_BOARD_REVISION} ...)"
       " given without BOARD argument, please specify BOARD"
     )
   endif()
 
-  if(DEFINED BUILD_STR_BOARD_QUALIFIERS AND NOT BUILD_STR_BOARD)
+  if(DEFINED BUILD_STR_BOARD_QUALIFIERS AND NOT BUILD_STR_BOARD AND NOT BUILD_STR_SOC)
     message(FATAL_ERROR
       "zephyr_build_string(${ARGV0} <list> BOARD_QUALIFIERS ${BUILD_STR_BOARD_QUALIFIERS} ...)"
-      " given without BOARD argument, please specify BOARD"
+      " given without BOARD or SOC argument, please specify BOARD or SOC"
+    )
+  endif()
+
+  if(DEFINED BUILD_STR_BOARD_REVISION AND BUILD_STR_SOC)
+    message(FATAL_ERROR
+      "zephyr_build_string(${ARGV0} <list> BOARD_REVISION ${BUILD_STR_BOARD_REVISION} ...)"
+      " given with SOC argument, these cannot be used together"
+    )
+  endif()
+
+  if(DEFINED BUILD_STR_SHORT AND BUILD_STR_SOC)
+    message(FATAL_ERROR
+      "zephyr_build_string(${ARGV0} <list> SHORT ${BUILD_STR_SHORT} ...)"
+      " given with SOC argument, these cannot be used together"
+    )
+  endif()
+
+  if(BUILD_STR_SOC AND NOT BUILD_STR_MERGE)
+    message(FATAL_ERROR
+      "zephyr_build_string(${ARGV0} <list> SOC ...)"
+      " given without MERGE argument, this must be used with the SOC option"
```

Comment:
MERGE should not be a requirement together with the SOC / BOARD_QUALIFIERS argument.

Keep zephyr_build_string generalized and let the caller decide if MERGE should be applied.
Note: see also other comment related to `MERGE`.

Reply:
Removed